### PR TITLE
Fix IPv6 non-specified ranges unexpectedly allowed

### LINKED ISSUE #1821: Problem: limit administrator functionality by IP address – add IPv6 support
### Current Behavior

**Steps to reproduce the behavior**

<ins>Setup</ins>

1. Browse to the AtoM site
2. Click the Login button and enter the email address and password of a user with the administrator privilege
3. Browse to the Security Settings page (`/index.php/settings/security`) and enter an IPv6 IP range into the box (separate the start and end range with a `–` or `,` character)

<ins>Reproduce</ins>

1. From an IPv6 machine outside of the IP range just added, launch a browser and browse to the AtoM site. Click the Login button and enter the email address and password of a user with the administrator privilege.
2. Click the Admin cog, and attempt to browse to any of the menu links

Current behaviour is that when attempting to access pages from an IPv6  address outside of the set ranges, AtoM allows access when the expectation is that it shouldn’t. This apparently happens when the client IP address is an IPv6 address. This unexpected behaviour does appear to be dependant on having entered an IPv6 range as a range restriction.

### Expected Behavior

When attempting to access pages from an IP address outside of the set ranges, AtoM should consistently return a **Sorry you do not have permission to access this page** message, irrespective of whether the remote address is IPv6 or IPv4.

### Possible Solution

Well, I'm not a PHP programmer, but I've managed to re-work `lib/filter/QubitLimitIp.class.php` to deliver the expected behaviour:
```
<?php

/*
 * This file is part of the Access to Memory (AtoM) software.
 *
 * Access to Memory (AtoM) is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * Access to Memory (AtoM) is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with Access to Memory (AtoM).  If not, see <http://www.gnu.org/licenses/>.
 */

class QubitLimitIpFilter extends sfFilter
{
    public function execute($filterChain)
    {
        $this->context = $this->getContext();
        $this->request = $this->context->getRequest();

        $this->limit = explode(';', sfConfig::get('app_limit_admin_ip'));

        // Pass if:
        // - Debug mode is on
        // - Setting "limit_admin_ip" is not set
        // - The filter is forwarding to admin/secure (isFirstCall)
        // - Route is user/logout
        if (
            $this->context->getConfiguration()->isDebug()
            || !$this->limit
            || !$this->isFirstCall()
            || ('user' == $this->request->getParameter('module') && 'logout' == $this->request->getParameter('action'))
        ) {
            $filterChain->execute();

            return;
        }

        // Forward to admin/secure if not allowed (only applies if user is authenticated)
        if ($this->context->user->isAuthenticated() && !$this->isAllowed()) {
            $this->context->getController()->forward(sfConfig::get('sf_secure_module'), sfConfig::get('sf_secure_action'));

            throw new sfStopException();
        }

        $filterChain->execute();
    }

    protected function getRemoteAddress()
    {
        $pathInfo = $this->request->getPathInfoArray();

        return $pathInfo['REMOTE_ADDR'];
    }

    protected function isAllowed()
    {
        $address = $this->getRemoteAddress();
        $addressBinary = inet_pton($address);

        // Check if empty
        if (1 == count($this->limit) && empty($this->limit[0])) {
            return true;
        }

        foreach ($this->limit as $item) {
            // Ranges are supported, using a comma or a dash
            $limit = preg_split('/[,-]/', $item);
            $limitBinary = inet_pton(trim($limit[0]));

            // Single IP
            if (1 == count($limit) && $addressBinary == $limitBinary && strlen($addressBinary) == strlen($limitBinary)) {
                return true;
            }

            // Range
            if (2 == count($limit)) {
                $limit[0] = trim($limit[0]);
                $limit[1] = trim($limit[1]);
                $firstInRangeBinary = inet_pton($limit[0]);
                $lastInRangeBinary = inet_pton($limit[1]);

                if (
                    (strlen($addressBinary) == strlen($firstInRangeBinary))
                     && ($addressBinary >= $firstInRangeBinary && $addressBinary <= $lastInRangeBinary)
                   ) {
                    return true;
                }
            }
        }
        return false;
    }

}
```
My changes replace `ip2long` with `inet_pton` as the latter can handle both IPv4 and IPv6 addresses. I've only done a very limited amount of testing, and would appreciate a second opinion. I have also added `strlen` checks, as according to [this SO answer](https://stackoverflow.com/a/14362786/1754517) "_when comparing them the length of the string must also be checked in order to prevent an IPv4 address to match the first part of an IPv6 address._".

### Context and Notes

A bit more local information and background:

- We are using the LIMIT ADMINISTRATOR FUNCTIONALITY BY IP ADDRESS feature [documented here](https://www.accesstomemory.org/en/docs/2.7/user-manual/administer/settings/#limit-administrator-functionality-by-ip-address)
- We have three “allowed” IP ranges set – two IPv4 ranges and one IPv6 range — on the assumption that both v4 and v6 are supported by the feature.
- I am using Opera browser's built-in VPN to test this. Changing the continent in the browser changes the IP address and sometimes the IP address type (v4/v6).
- From the Nginx access logs (`/var/log/nginx/access.log`), it appears that when the client IP address is IPv4, AtoM behaves as expected – access is blocked (403 response code). However, when the client IP address is IPv6, access is wrongly allowed (200 response code)
- Sample log lines below:

```
77.111.245.12 - - [20/May/2024:14:57:14 +0100] "GET /index.php/settings/global HTTP/2.0" 403 4331 "https://borthcat.york.ac.uk/index.php/accession/browse" "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/123.0.0.0 Safari/537.36 OPR/109.0.0.0"
2001:67c:2628:647:f::223 - - [20/May/2024:15:03:46 +0100] "GET /index.php/settings/global HTTP/2.0" 200 7210 "https://borthcat.york.ac.uk/index.php/aclGroup/list" "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/123.0.0.0 Safari/537.36 OPR/109.0.0.0"
```
It looks like `lib/filter/QubitLimitIp.class.php` is handling this behind the scenes. This file uses the [ip2long function](https://www.php.net/manual/en/function.ip2long.php), which is apparently only designed to cope with IPv4 IP addresses (the ip2long function is described:  “_Converts a string containing an (IPv4) Internet Protocol dotted address into a long integer_”).

### Version used

AtoM 2.7.3

### Operating System and version

Ubuntu 20.04

### Default installation culture

en

### PHP version

PHP 7.4

### Contact details

_No response_

### PR DESCRIPTION
**Pull request by Jimadine**
As reported in https://github.com/artefactual/atom/issues/1821.

As alluded to in the issue text, I am entirely happy for someone to pull my changes apart and rewrite. We are already using the modified `QubitLimitIp.class.php` in production, and it meets our basic needs.

For the future: it would be a good idea to also validate the input box content so that the administrator adding IPs/ranges can be sure what they have entered meets the required format. I've created [a simple example](https://www-users.york.ac.uk/~jea9/ipaddressvalidation.html) of how this could be done in Javascript, but it could be implemented differently e.g. entirely server-side.

### CODE REVIEW

**Comment by anvit** [comment]
Hi @Jimadine ! Thank you so much for both, reporting this issue and submitting this PR to fix it. We'll review the PR soon, and if everything looks good, get this fix in for the next release.

When you get a chance, would you be able to sign our Contributor Agreement? Here's a link with details about it agreement —
https://wiki.accesstomemory.org/Development/Contribute_code#Copyright_and_license

**Comment by Jimadine** [comment]
> Hi @Jimadine ! Thank you so much for both, reporting this issue and submitting this PR to fix it. We'll review the PR soon, and if everything looks good, get this fix in for the next release.

You're very welcome! 😄 

> When you get a chance, would you be able to sign our Contributor Agreement? Here's a link with details about it agreement — https://wiki.accesstomemory.org/Development/Contribute_code#Copyright_and_license

I've just this moment emailed my signed agreement to agreement@artefactual.com

**Comment by anvit** [comment]
@Jimadine yup, we received it. Thanks!

**Comment by anvit** [comment]
@Jimadine I just looked at your changes and this fix makes sense to me. I noticed that our syntax checks are failing (probably indentation related), and we like to keep our commit message header to 50 characters but the rest looks good! Would you be able to fix those? If not, I'm happy to add a commit fixing those and then accept the changes!

**Comment by Jimadine** [comment]
> @Jimadine I just looked at your changes and this fix makes sense to me. I noticed that our syntax checks are failing (probably indentation related), and we like to keep our commit message header to 50 characters but the rest looks good! Would you be able to fix those? If not, I'm happy to add a commit fixing those and then accept the changes!

Hi @anvit - I have amended the commit message and fixed the indentation. Hopefully the syntax checks will pass now 😃.

File: lib/filter/QubitLimitIp.class.php [hunk 1]
```diff
@@ -71,28 +72,29 @@ protected function isAllowed()
...
                 if (
-                    ip2long($limit[0]) <= $addressLong
-                    && ip2long($limit[1]) >= $addressLong
+                    (strlen($addressBinary) == strlen($firstInRangeBinary))
+                     && ($addressBinary >= $firstInRangeBinary && $addressBinary <= $lastInRangeBinary)
                 ) {
                     return true;
                 }
             }
         }
```
Comment (anvit) [inline lib/filter/QubitLimitIp.class.php]: Syntax checks seem to be failing due to missing new line before return statement. Please add an empty new line before the return here.

File: lib/filter/QubitLimitIp.class.php [hunk 2]
```diff
@@ -71,28 +72,29 @@ protected function isAllowed()
...
-
         return false;
     }
+
```
Comment (anvit) [inline lib/filter/QubitLimitIp.class.php]: Syntax checker also asks to remove this empty line.

**Review by anvit — CHANGES_REQUESTED** [review]
@Jimadine unfortunately syntax checks are still failing. I found the lines it was complaining about and flagged them in the review.

**Comment by Jimadine** [comment]
@anvit - thank you for identifying what needs to be changed 😄. When I ran PHP CS fixer with the default rules, it made the following change:

```
diff --git a/lib/filter/QubitLimitIp.class.php b/lib/filter/QubitLimitIp.class.php
index 41b32ed06..ff053f4db 100644
--- a/lib/filter/QubitLimitIp.class.php
+++ b/lib/filter/QubitLimitIp.class.php
@@ -89,7 +89,7 @@ class QubitLimitIpFilter extends sfFilter
                 if (
                     (strlen($addressBinary) == strlen($firstInRangeBinary))
                      && ($addressBinary >= $firstInRangeBinary && $addressBinary <= $lastInRangeBinary)
-                   ) {
+                ) {
                     return true;
                 }
             }
```

But it didn't make the changes you've identified. I'm wondering if that's because there are customisations to the rules. I can see a rule in `.php-cs-fixer.dist.php`, in the root of the repo, but even when I reference this file with `--cache-file`, I get the same result. This is the output I get:
```
25/06/2024  9:15:54> - C:\Users\jea9\AppData\Local\Programs\php-8.3.8-nts-Win32-vs16-x64>php php-cs-fixer.phar fix C:\Jims_file_repository\Version_control\GitHub\atom-york\lib\filter\QubitLimitIp.class.php --cache-file C:\Jims_file_repository\Version_control\GitHub\atom-york\.php-cs-fixer.dist.php
PHP CS Fixer 3.59.3 (064efa1) 7th Gear by Fabien Potencier, Dariusz Ruminski and contributors.
PHP runtime: 8.3.8
Running analysis on 1 core sequentially.
You can enable parallel runner and speed up the analysis! Please see usage docs for more information.
Loaded config default.
Using cache file "C:\Jims_file_repository\Version_control\GitHub\atom-york\.php-cs-fixer.dist.php".
 1/1 [...] 100%


Fixed 0 of 1 files in 0.004 seconds, 20.00 MB memory used
```
Do you have any thoughts on why PHP CS fixer isn't identifying/fixing those lines for me? No worries if not; I can make the changes manually.

**Comment by anvit** [comment]
@Jimadine Hmm… I'm not entirely sure but my guess is that the rules aren't being picked up? I usually run it via the composer script like this `composer php-cs-fix -- fix -v`. There's another thing I noticed. It says you're running `php 8.3.8`. Do you usually do all the development work for AtoM directly on your machine? Or do you work through a docker container? Because AtoM uses an older versions of PHP (7.4) and composer (3.6.0), so I wonder if either of those are factors as well. If you're using docker, you can run it using `docker compose exec atom composer php-cs-fix -- fix -v`

**Comment by Jimadine** [comment]
@anvit - I don't usually do AtoM development, so I downloaded the latest version of PHP and php-cs-fixer.phar to a folder on my Windows machine. My testing has been on our Ubuntu 20.04 VMWare test server. I tend to use Vagrant a lot, though more for deployment testing purposes.

For this purpose, I've spun up a Vagrant Ubuntu 20.04 box with PHP 7.4, installed the composer dependencies with `php composer.phar update`, and tried again to see what PHP CS fixer will fix. This time it made the following corrections (on top of the previous undentation):
```
root@ubuntu-focal:/usr/share/nginx/atom# php composer.phar php-cs-fixer fix /usr/share/nginx/atom/lib/filter/QubitLimitIp.class.php
Do not run Composer as root/super user! See https://getcomposer.org/root for details
Continue as root/super user [yes]? yes
> php-cs-fixer 'fix' '/usr/share/nginx/atom/lib/filter/QubitLimitIp.class.php'
PHP CS Fixer 3.59.3 7th Gear by Fabien Potencier, Dariusz Ruminski and contributors.
PHP runtime: 7.4.3-4ubuntu2.23
Running analysis on 1 core sequentially.
You can enable parallel runner and speed up the analysis! Please see usage docs for more information.
Loaded config default from "/usr/share/nginx/atom/.php-cs-fixer.dist.php".
Using cache file ".php-cs-fixer.cache".
Paths from configuration file have been overridden by paths provided as command arguments.
 1/1 [...] 100%

   1) lib/filter/QubitLimitIp.class.php

Fixed 1 of 1 files in 0.142 seconds, 18.00 MB memory used
root@ubuntu-focal:/usr/share/nginx/atom# git diff lib/filter/QubitLimitIp.class.php
diff --git a/lib/filter/QubitLimitIp.class.php b/lib/filter/QubitLimitIp.class.php
index ff053f4..cb0dbf6 100644
--- a/lib/filter/QubitLimitIp.class.php
+++ b/lib/filter/QubitLimitIp.class.php
@@ -94,7 +94,7 @@ class QubitLimitIpFilter extends sfFilter
                 }
             }
         }
+
         return false;
     }
-
 }
```
So in addition to removing line 99 that you identified, it wants a new line at line 97, before the `return false;`. I've made these changes, and I've also added the extra new line at line 93, as you mentioned.

I'm not sure why I'm seeing these discrepancies. I have installed the latest Composer version `2.7.7 2024-06-10 22:11:12`. When I run `php composer.phar php-cs-fixer -- fix` it fixes _lots_ of files.

I see the syntax checking has failed on my fork again 😞.

File: lib/filter/QubitLimitIp.class.php (hunk 1 above)
Comment (Jimadine) [reply lib/filter/QubitLimitIp.class.php]: After adding the extra blank line, CS fixer complained about it:
```
root@ubuntu-focal:/usr/share/nginx/atom# php composer.phar php-cs-fixer fix /usr/share/nginx/atom/lib/filter/QubitLimitIp.class.php -v
Do not run Composer as root/super user! See https://getcomposer.org/root for details
Continue as root/super user [yes]?
> php-cs-fixer: php-cs-fixer 'fix' '/usr/share/nginx/atom/lib/filter/QubitLimitIp.class.php'
PHP CS Fixer 3.59.3 7th Gear by Fabien Potencier, Dariusz Ruminski and contributors.
PHP runtime: 7.4.3-4ubuntu2.23
Running analysis on 1 core sequentially.
You can enable parallel runner and speed up the analysis! Please see usage docs for more information.
Loaded config default from "/usr/share/nginx/atom/.php-cs-fixer.dist.php".
Using cache file ".php-cs-fixer.cache".
Paths from configuration file have been overridden by paths provided as command arguments.
 1/1 [...] 100%

   1) lib/filter/QubitLimitIp.class.php (no_extra_blank_lines)

Fixed 1 of 1 files in 0.125 seconds, 18.00 MB memory used
root@ubuntu-focal:/usr/share/nginx/atom# git diff lib/filter/QubitLimitIp.class.php
diff --git a/lib/filter/QubitLimitIp.class.php b/lib/filter/QubitLimitIp.class.php
index 592277d..cb0dbf6 100644
--- a/lib/filter/QubitLimitIp.class.php
+++ b/lib/filter/QubitLimitIp.class.php
@@ -90,7 +90,6 @@ class QubitLimitIpFilter extends sfFilter
                     (strlen($addressBinary) == strlen($firstInRangeBinary))
                      && ($addressBinary >= $firstInRangeBinary && $addressBinary <= $lastInRangeBinary)
                 ) {
-
                     return true;
                 }
             }
```
...so I went with the CS fixer fix to remove it, and now the syntax check passes 🚀

**Comment by Jimadine** [comment]
@anvit - I think all should be well now - [see my comment](https://github.com/artefactual/atom/pull/1830#discussion_r1655158639).

One for a new issue perhaps, but it might be helpful to add the `-v` option to the PHP CS fixer command in the Github workflow. Doing this appears to give a description of what the problem is, in the output, e.g.:
```
   1) lib/filter/QubitLimitIp.class.php (no_extra_blank_lines)
```
It's not ideal that there's no line number reference, but at least it's something.

**Review by anvit — APPROVED** [review]
@Jimadine Thank you! Your last commit fixed the syntax check issues!

**Pull request merged**